File: 2020-03/3-14/45/45.py


from typing import List


class Solution:
    def jump(self, nums: List[int]) -> int:

        end = 0
        max_pos = 0
        count = 0

        for i in range(len(nums) - 1):
            max_pos = max(max_pos, i + nums[i])

            if i == end:
                # 第i+1次的踏板范围全部跑完，才会更新一次count
                end = max_pos
                count += 1

            # end到终点，提前结束
            if end == len(nums) - 1:
                break

        return count

File: 2020-05/5-04/45跳跃游戏II/45.cpp


class Solution {
public:
    int jump(vector<int>& nums) {
        int start = 0, end = 0; // 左闭右闭区间
        int time = 0;

        while(end < nums.size() - 1){
            int max_pos = 0;
            for(int i = start; i <= end; i++){
                max_pos = max(max_pos, i + nums[i]);
                if(max_pos == nums.size()- 1)  // 提前终止一下
                    return time + 1;
            }
            start = end + 1;
            end = max_pos;
            time++;
        }
        return time;
    }
};